src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs

```diff
@@ -1285,43 +1256,36 @@ protected override NamedPipeClientStream DoConnect(int timeout)
             //
             int startTime = Environment.TickCount;
             int elapsedTime = 0;
-            SafePipeHandle pipeHandle = null;
+            SafeFileHandle sf = null;
 
             do
             {
-                // Get handle to pipe.
-                pipeHandle = NamedPipeNative.CreateFile(
-                    lpFileName: PipeName,
-                    dwDesiredAccess: NamedPipeNative.GENERIC_READ | NamedPipeNative.GENERIC_WRITE,
-                    dwShareMode: 0,
-                    SecurityAttributes: IntPtr.Zero,
-                    dwCreationDisposition: NamedPipeNative.OPEN_EXISTING,
-                    dwFlagsAndAttributes: NamedPipeNative.FILE_FLAG_OVERLAPPED,
-                    hTemplateFile: IntPtr.Zero);
-
-                int lastError = Marshal.GetLastWin32Error();
-                if (pipeHandle.IsInvalid)
+                try
                 {
-                    if (lastError == NamedPipeNative.ERROR_FILE_NOT_FOUND)
-                    {
-                        elapsedTime = unchecked(Environment.TickCount - startTime);
-                        Thread.Sleep(100);
-                        continue;
-                    }
-                    else
-                    {
-                        throw new PSInvalidOperationException(
-                            StringUtil.Format(RemotingErrorIdStrings.CannotConnectContainerNamedPipe, lastError));
-                    }
+                    sf = File.OpenHandle(PipeName, FileMode.Open, FileAccess.ReadWrite, FileShare.None, FileOptions.Asynchronous);
+                }
+                catch (FileNotFoundException)
+                {
+                    elapsedTime = unchecked(Environment.TickCount - startTime);
+                    Thread.Sleep(100);
+                    continue;
                 }
-                else
+                catch (Exception exc)
                 {
-                    break;
+                    throw new PSInvalidOperationException(StringUtil.Format(RemotingErrorIdStrings.CannotConnectContainerNamedPipe, exc.HResult), exc);
                 }
+
+                break;
             } while (elapsedTime < timeout);
 
+            ArgumentNullException.ThrowIfNull(sf);
+
+            bool gotRef = false;
+            SafePipeHandle pipeHandle = null;
             try
             {
+                sf.DangerousAddRef(ref gotRef);
```

Comment:
Not sure if this is necessary as the safe handle is passed to `pipehandle` object.

Reply:
We need to re-pack SafeFileHandle to SafePipeHandle. Since SafeFileHandle is disposable it can close the handle and we must block this.